nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,617 @@
+    if (polygon->getActionType() == STOP) {
+      // Setting up zero velocity for STOP model
+      robot_action.action_type = STOP;
+      robot_action.req_vel.x = 0.0;
+      robot_action.req_vel.y = 0.0;
+      robot_action.req_vel.tw = 0.0;
+      return true;
+    } else {  // SLOWDOWN
+      const Velocity safe_vel = velocity * polygon->getSlowdownRatio();
+      // Check that currently calculated velocity is safer than
+      // chosen for previous shapes one
+      if (safe_vel < robot_action.req_vel) {
+        robot_action.action_type = SLOWDOWN;
+        robot_action.req_vel = safe_vel;
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+bool CollisionMonitor::processApproach(
+  const std::vector<Point> & collision_points,
+  const Velocity & velocity,
+  Action & robot_action) const
+{
+  if (footprint_sub_ != nullptr) {
+    // Get robot footprint from footprint subscriber
+    std::vector<geometry_msgs::msg::Point> footprint_vec;
+    std_msgs::msg::Header footprint_header;
+    footprint_sub_->getFootprintInRobotFrame(footprint_vec, footprint_header);
+
+    // Update footprint polygon points with latest received message
+    footprint_->setPolygon(footprint_vec);
+  }
+
+  // Obtain time before a collision
+  const double collision_time = getCollisionTime(collision_points, velocity);
+  if (collision_time >= 0.0)
+  {
+    // If collision will occurr, reduce robot speed
+    const double change_ratio = collision_time / time_before_collision_;
+    const Velocity safe_vel = velocity * change_ratio;
+    // Check that currently calculated velocity is safer than
+    // chosen for previous shapes one
+    if (safe_vel < robot_action.req_vel) {
+      robot_action.action_type = APPROACH;
+      robot_action.req_vel = safe_vel;
+      return true;
+    }
+  }
+
+  return false;
+}
+
+double CollisionMonitor::getCollisionTime(
+  const std::vector<Point> & collision_points,
+  const Velocity & velocity) const
+{
+  // Initial robot pose is {0,0} in base_footprint coordinates
+  Pose pose = {0.0, 0.0, 0.0};
+  Velocity vel = velocity;
+
+  // Array of points transformed to the frame concerned with pose on each simulation step
+  std::vector<Point> points_transformed;
+
+  // Robot movement simulation
+  for (double time = 0.0; time <= time_before_collision_; time += simulation_time_step_) {
+    // Shift the robot pose towards to the vel during simulation_time_step_ time interval
+    // NOTE: vel is changing during the simulation
+    projectState(simulation_time_step_, pose, vel);
+    // Transform collision_points to the frame concerned with current robot pose
+    points_transformed = collision_points;
+    transformPoints(pose, points_transformed);
+    // If the collision occurred on this stage, return the actual time before a collision
+    // as if robot was moved with given velocity
+    if (footprint_->getPointsInside(points_transformed) > footprint_->getMaxPoints()) {
```

Comment:
I don't understand how to read this. Why does a footprint contain the maximum number of valid points?